tests/imagecheck.py

```diff
@@ -0,0 +1,193 @@
+            had_error = True
+    return had_error
+
+
+def print_test_error(cmd, expect_text, cmd_output):
+    print("Error running {}".format(cmd))
+    print("Expected Regex: {}".format(expect_text))
+    print("Actual: {}".format(cmd_output))
+
+
+def find_unittest_info():
+    """
+    Recursively walk the current directory looking for tests.
+    :return: [(test_dir, imagename, test_filename)]: list of testing data tuples
+    """
+    test_info = []
+    for root, dirs, files in os.walk("."):
+        for name in files:
+            if name == UNITTEST_FILENAME:
+                test_directory = root
+                imagename = "localhost:5000/test_{}".format(test_directory.replace("./", "").replace("/", ":").replace("+","_"))
+                test_filename = os.path.join(root, name)
+                test_info.append((test_directory, imagename, test_filename))
+    return test_info
+
+
+def get_test_file_path(file_path):
+    """
+    Given a file path return a path to the associated unittest.yml file.
+    This is either the file_path itself if it points to unittest.yml or if a Dockerfile it will
+    point to the unittest.yml file(if it exists)
+    :param file_path: str: path to a changed file that may have an associated unittest.yml file
+    :return: str: a path to a unittest.yml file to test or None if there is no associated test file
+    """
+    filename = os.path.basename(file_path)
+    if filename in ["Dockerfile", UNITTEST_FILENAME]:
+        parent_directory = os.path.dirname(file_path)
+        unittest_filepath = "{}/{}".format(parent_directory, UNITTEST_FILENAME)
+        if os.path.exists(unittest_filepath):
+            return unittest_filepath
+    return None
+
+
+def get_unittest_file_paths(path_list):
+    """
+    For a list of paths return any associated unittest.yml files.
+    :param path_list: [str]: list of paths
+    :return: [str]: list of unitttest.yml paths
+    """
+    unittest_paths = set()
+    for file_path in path_list:
+        unittest_path = get_test_file_path(file_path)
+        if unittest_path:
+            unittest_paths.add(unittest_path)
+    return unittest_paths
+
+
+def find_and_run_tests(owner, changed_paths):
+    """
+    Find an run tests based on a docker ownername (used to build image name) and a list of paths
+    :param owner: str: prefix of docker imagename
+    :param changed_paths: [str]: list of paths that were changed and may need to be tested
+    :return: bool: true when we had errors
+    """
+    had_errors = False
+    tested_images = 0
+    images_with_errors = 0
+    for unittest_path in get_unittest_file_paths(changed_paths):
+        parts = unittest_path.split(sep="/")
+        if len(parts):
+            tool, tag, _ = parts
+            imagename = "{}/{}:{}".format(owner, tool, tag).replace("+","_")
+            build_docker_image(imagename, "{}/{}".format(tool, tag))
```

Comment:
Since `build-ci.sh` is designed to build images, it's unexpected that the test step has separate logic for determining what to build and how to build it.

There's also a risk if that logic is not maintained identically. Build and test can produce different images. Deploy, coming after test, would be pushing the images built by the test step, and not the ones from build.

I recognize that test needs to detect changes to the unittest.yml files for re-running tests, but I'm not sure those need to trigger a rebuild.

Reply:
Looks like we need the `+`->`_` substitution in `functions.sh` too. Ideally that logic would be in a single place, but for now, just a note that my `functions.sh` is broken 😃

Reply:
Created #45 to handle `+` to `_` issue.